drivers/ak8963/AK8963.cpp

```diff
@@ -0,0 +1,430 @@
+	    && in_range(ak8963_report.y * _mag_sens_adj[1], -200.0, 200.0)
+	    && in_range(ak8963_report.z * _mag_sens_adj[2], -3200.0, -800.0)) {
+		DF_LOG_INFO("Selftest passed!");
+		passed_test = 0;
+
+	} else {
+		DF_LOG_ERR("Selftest failed!");
+		passed_test = -1;
+	}
+
+	if (result != 0) {
+		DF_LOG_ERR("Error reading data");
+		return -1;
+	}
+
+	bits = AK8963_BITS_ASTC_NORMAL;
+
+	if (_writeReg(AK8963_REG_ASTC, &bits, 1) != 0) {
+		DF_LOG_ERR("Test: Set self test");
+		return -1;
+	}
+
+	usleep(1000);
+	bits = AK8963_BITS_CNTL1_MODE_POWER_DOWN;
+
+	if (_writeReg(AK8963_REG_CNTL1, &bits, 1) != 0) {
+		DF_LOG_ERR("Test: Set power down mode");
+		return -1;
+	}
+
+	usleep(1000);
+	return passed_test;
+}
+
+int AK8963::ak8963_init()
+{
+
+	/* Zero the struct */
+	m_synchronize.lock();
+
+	m_sensor_data.last_read_time_usec = 0;
+	m_sensor_data.read_counter = 0;
+	m_sensor_data.error_counter = 0;
+	m_synchronize.unlock();
+
+	// Perform soft-reset
+	uint8_t bits = AK8963_BITS_CNTL2_SOFT_RESET;;
+	int result = _writeReg(AK8963_REG_CNTL2, &bits, 1);
+
+	if (result < 0) {
+		DF_LOG_ERR("AK8963 soft reset failed.");
+		return -1;
+	}
+
+	usleep(1000);
+
+	// Detect mag presence by reading whoami register
+	if (detect() != 0) {
+		DF_LOG_ERR("AK8963 mag not detected.");
+		return -1;
+	}
+
+	// Get mag calibraion data from Fuse ROM
+	if (get_sensitivity_adjustment() != 0) {
+		DF_LOG_ERR("Unable to read mag sensitivity adjustment");
+		return -1;
+	}
+
+	// Power on and configure the mag to produce 16 bit data in continuous measurement mode.
+	bits = AK8963_BITS_CNTL1_OUTPUT_16BIT | AK8963_BITS_CNTL1_MODE_CONTINOUS2;
+	result = _writeReg(AK8963_REG_CNTL1, &bits, 1);
+
+	if (result != 0) {
+		DF_LOG_ERR("Unable to configure the magnetometer mode.");
+	}
+
+	usleep(1000);
+	return 0;
+}
+
+int AK8963::start()
+{
+	int result = I2CDevObj::start();
+
+	if (result != 0) {
+		DF_LOG_ERR("error: could not start DevObj");
+		goto exit;
+	}
+
+	/* Configure the I2C bus parameters for the mag sensor. */
+	result = _setSlaveConfig(AK8963_SLAVE_ADDRESS,
+				 AK8963_BUS_FREQUENCY_IN_KHZ,
+				 AK8963_TRANSFER_TIMEOUT_IN_USECS);
+
+	if (result != 0) {
+		DF_LOG_ERR("I2C slave configuration failed");
+		goto exit;
+	}
+
+	/* Initialize the mag sensor. */
+	result = ak8963_init();
+
+	if (result != 0) {
+		DF_LOG_ERR("error: mag sensor initialization failed, sensor read thread not started");
+		goto exit;
+	}
+
+
+	result = DevObj::start();
+
+	if (result != 0) {
+		DF_LOG_ERR("error: could not start DevObj");
+		goto exit;
+	}
+
+exit:
+	return result;
+}
+
+int AK8963::stop()
+{
+	int result = DevObj::stop();
+
+	if (result != 0) {
+		DF_LOG_ERR("DevObj stop failed");
+		return result;
+	}
+
+	// Leave in a power-down mode
+	uint8_t bits = AK8963_BITS_CNTL1_MODE_POWER_DOWN;
+
+	if (_writeReg(AK8963_REG_CNTL1, &bits, 1) != 0) {
+		DF_LOG_ERR("Sensitivity: Set power down mode");
+		return -1;
+	}
+
+	usleep(100000);
```

Comment:
I wait a short-time after changing the mode, making sure there is no new request. But it also should happen before `DevObj::stop()` and not after.